archive/imo/imo2005_q3.lean

```diff
@@ -20,21 +21,17 @@ and then making use of `xyz ≥ 1` to show `(x^5-x^2)/(x^3*(x^2+y^2+z^2)) ≥ (x
 lemma key_insight (x y z : ℝ) (hx : x > 0) (hy : y > 0) (hz : z > 0) (h : x*y*z ≥ 1) :
   (x^5-x^2)/(x^5+y^2+z^2) ≥ (x^2-y*z)/(x^2+y^2+z^2) :=
 begin
-  have h₁ : 0 < x^5+y^2+z^2, linarith [pow_pos hx 5, pow_pos hy 2, pow_pos hz 2],
-  have h₂ : 0 < x^3, exact pow_pos hx 3,
-  have h₃ : 0 < x^2+y^2+z^2, linarith [pow_pos hx 2, pow_pos hy 2, pow_pos hz 2],
-  have h₄ : 0 < x^3*(x^2+y^2+z^2), exact mul_pos h₂ h₃,
+  have h₁ : 0 < x^5+y^2+z^2 := by positivity,
+  have h₂ : 0 < x^3 := by positivity,
+  have h₃ : 0 < x^2+y^2+z^2 := by positivity,
+  have h₄ : 0 < x^3*(x^2+y^2+z^2) := by positivity,
 
   have key : (x^5-x^2)/(x^5+y^2+z^2) - (x^5-x^2)/(x^3*(x^2+y^2+z^2))
            = ((x^3 - 1)^2*x^2*(y^2 + z^2))/((x^5+y^2+z^2)*(x^3*(x^2+y^2+z^2))),
   { field_simp [h₁.ne', h₄.ne'],
```

Comment:
Would it be worth to change `field_simp` to try to discharge its nonzero goals by trying to apply `positivity`? Something like replacing lines 20--25 of `tactic/field_simp.lean` with
```
meta def field_simp.ne_zero : tactic unit := do
  goal ← tactic.target,
  match goal with
  | `(%%e ≠ 0) := assumption <|> (do n ← e.to_rat, `[norm_num1])
      <|> `[apply ne_of_gt, positivity] <|> `[apply ne_of_lt, positivity]
  | _ := tactic.fail "goal should be of the form `x ≠ 0`"
  end
```

Reply:
We are thinking along the same lines!  I hope that `positivity` will eventually be a default discharger for several tactics (`field_simp` with `ne_of_lt`, general `simp`, `continuity`, an experimental "inequality-congr" I've been writing).

But the current version of `positivity` is limited, and I think it might be premature to build it into `field_simp` now.  If you don't mind, I'd rather wait a few months and see how often `positivity` is useful in the wild when people are setting up `field_simp`s.

Reply:
ok. Maybe I'll make some little experiments on my side, but not in this PR.

Reply:
I'd be very interested in your experiments!